src/Storages/MergeTree/MergeTreeRangeReader.cpp

```diff
@@ -1313,6 +1322,56 @@ static ColumnPtr combineFilters(ColumnPtr first, ColumnPtr second)
     return mut_first;
 }
 
+void MergeTreeRangeReader::executeActionsForReadHints(ReadResult & result) const
+{
+    auto before_rows = result.num_rows;
+
+    /// Populate a filter that is True only for the exact "neighbour" part offsets we got from vector index
+    auto filter_data = ColumnUInt8::create(result.num_rows, UInt8(0));
+    IColumn::Filter & filter = filter_data->getData();
+
+    /// Populate the "_distance" virtual column from the distances we got from vector index
+    auto distance_pos = read_sample_block.getPositionByName("_distance");
+    auto distance_column = IColumn::mutate(std::move(result.columns[distance_pos]));
+    auto & distances = typeid_cast<ColumnFloat32 *>(distance_column.get())->getData();
+
+    const auto & read_hints = merge_tree_reader->data_part_info_for_read->getReadHints();
+    const auto & offsets_and_distances = read_hints.ann_search_results.value();
+    auto exact_part_offsets = offsets_and_distances.first;
```

Comment:
The row numbers returned by vector similarity indexes are per index granule (e.g. 2x the size of a normal granule in index definition `INDEX idx vec TYPE vector_similarity('hnsw', 'L2Distance', 2) GRANULARITY 2`). A part may have many index granules. I am slightly confused why the variable is named "exact_part_offset".